synapse/lib/aha.py

```diff
@@ -537,41 +683,100 @@ def _initCellHttpApis(self):
+            # bootstrap a CA for our aha:network
+            netw = self.conf.req('aha:network')
 
-                name = self.conf.get('aha:name')
+            if self.certdir.getCaCertPath(netw) is None:
+                logger.info(f'Adding CA certificate for {netw}')
+                await self.genCaCert(netw)
 
+            name = self.conf.get('aha:name')
+            if name is not None:
                 host = f'{name}.{netw}'
                 if self.certdir.getHostCertPath(host) is None:
                     logger.info(f'Adding server certificate for {host}')
                     await self._genHostCert(host, signas=netw)
 
-                user = self._getAhaAdmin()
-                if user is not None:
-                    if self.certdir.getUserCertPath(user) is None:
-                        logger.info(f'Adding user certificate for {user}@{netw}')
-                        await self._genUserCert(user, signas=netw)
+            root = f'root@{netw}'
+            await self._genUserCert(root, signas=netw)
+
+            user = self.conf.get('aha:admin')
+            if user is not None:
+                await self._genUserCert(user, signas=netw)
+
+    def _getDnsName(self):
+        # emulate the old aha name.network behavior if the
+        # explicit option is not set.
+
+        hostname = self.conf.get('dns:name')
+        if hostname is not None:
+            return hostname
+
+        ahaname = self.conf.get('aha:name')
+        ahanetw = self.conf.get('aha:network')
+        if ahaname is not None and ahanetw is not None:
+            return f'{ahaname}.{ahanetw}'
+
+    def _getProvListen(self):
+
+        lisn = self.conf.get('provision:listen')
+        if lisn is not None:
+            return lisn
+
+        # this may not use _getDnsName() in order to maintain
+        # backward compatibilty with aha name.network configs
+        # that do not intend to listen for provisioning.
+        hostname = self.conf.get('dns:name')
+        if hostname is not None:
+            return f'ssl://{hostname}:27272'
+
+    def _getDmonListen(self):
+
+        lisn = self.conf.get('dmon:listen', s_common.novalu)
+        if lisn is not s_common.novalu:
+            return lisn
+
+        network = self.conf.req('aha:network')
+        dnsname = self._getDnsName()
+        if dnsname is not None:
+            return f'ssl://0.0.0.0?hostname={dnsname}&ca={network}'
+
+    def _reqProvListen(self):
+        lisn = self._getProvListen()
+        if lisn is not None:
+            return lisn
+
+        mesg = 'The AHA server is not configured for provisioning.'
+        raise s_exc.NeedConfValu(mesg=mesg)
 
     async def initServiceNetwork(self):
 
+        # bootstrap CA/host certs first
+        network = self.conf.req('aha:network')
+
+        hostname = self._getDnsName()
+        if hostname is not None and network is not None:
+            await self._genHostCert(hostname, signas=network)
```

Comment:
Should mirrors be calling this and potentially generating new CA certs that would overwrite the CA cert on the leader?

Reply:
since clones/mirrors begin with a backup of an existing one, this will always be present. were you thinking this should only run if the AHA service is the leader?

Reply:
They should be present yeah, I'm just thinking of the case where a mirror starts up with the certs missing for some reason and it generates a new CA cert which would push to the leader and overwrite any existing cert. Seems like genCaCert should only run on the leader.

Reply:
Updated `_genHostCert()` to not automatically generate the CA cert